### PR TITLE
Verify Paid Newsletter Access level before checking User Subscriptions

### PR DESCRIPTION
**Pull request by romarioraffington**
## Background
There has been an overall increased connection to the DB and the hourly spike due to how we are gating comments. See `maybe_gate_existing_comments`. This `maybe_gate_existing_comments` filter is being called for all comments on WPCOM. To fix this, we first verify that a comment should be gated by checking the post access level. If the post should not be gated, we return early. 

The goal here is to not query the DB for every comment or post on WPCOM. We only query for posts/comments that could be potentially gated.

See p1684456374317729/1684412809.266309-slack-C02AVAR9B for more context. A quick fix was merged [here](https://github.com/Automattic/jetpack/pull/30355). This PR improves on that fix.

## Proposed changes:

* Verify Paid Newsletter access level should be gated before querying DB. 
* Replace newsletter references with post level. This can be confusing. See p1684457715342319/1684412809.266309-slack-C02AVAR9B
* Consolidate get_post_access_level functionality into the `Jetpack_Membership` class

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Does this pull request change what data or activity we track or use?

No

## Testing instructions:

1. Create a post with a newsletter access level
2. Verify the post is gated correctly
3. Repeat steps 1 -2 with the another newsletter access level
4. Add the Subscribe block to a post
5. Verify the Subscribe functionality works as expected with a "free" and "paid" plan

### CODE REVIEW

File: projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php [hunk 1]
```diff
@@ -788,7 +788,7 @@ function maybe_close_comments( $default_comments_open, $post_id ) {
  * @return string
  */
 function maybe_gate_existing_comments( $comment ) {
-	if ( empty( $comment ) ) {
+	if ( empty( $comment ) || 'everybody' === get_post_access_level() ) {
```
Comment (romarioraffington) [inline projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php]: I'm unsure if we want to query the post meta for every comment. Either way, it's cached, https://developer.wordpress.org/reference/functions/get_metadata_raw/


Here we use the string `everybody` because `Token_Subscription_Service` is not being loaded on Jetpack sites. See https://github.com/Automattic/jetpack/pull/30172#pullrequestreview-1430057320

**Comment by n3f** [comment]
We should get a WPCOM PR going to address the caching in `earn_get_user_subscriptions_for_site_id` and have some unit tests verify this is working correctly...

File: projects/plugins/jetpack/modules/subscriptions/views.php [hunk 2]
```diff
@@ -366,7 +347,7 @@ public static function render_widget_subscription_form( $args, $instance, $subsc
 		$submit_button_wrapper_styles = isset( $instance['submit_button_wrapper_styles'] ) ? $instance['submit_button_wrapper_styles'] : '';
 		$email_field_classes          = isset( $instance['email_field_classes'] ) ? $instance['email_field_classes'] : '';
 		$email_field_styles           = isset( $instance['email_field_styles'] ) ? $instance['email_field_styles'] : '';
-		$post_access_level            = self::get_post_access_level();
+		$post_access_level            = \Jetpack_Memberships::get_post_access_level();
```
Comment (romarioraffington) [inline projects/plugins/jetpack/modules/subscriptions/views.php:350]: Verify `Jetpack_Memberships` is available

File: projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php [hunk 3]
```diff
@@ -477,28 +477,6 @@ function render_block( $attributes ) {
 	return render_jetpack_subscribe_form( $data, $classes, $styles );
 }
 
-/**
- * Generates the source parameter to pass to the iframe
- *
- * @return string the actual post access level (see projects/plugins/jetpack/extensions/blocks/subscriptions/settings.js for the values).
- */
-function get_post_access_level() {
```
Comment (romarioraffington) [inline projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php:485]: Moved this to `Jetpack_Memberships` class

File: projects/plugins/jetpack/modules/memberships/class-jetpack-memberships.php [hunk 4]
```diff
@@ -39,7 +39,7 @@ class Jetpack_Memberships {
 	 *
 	 * @var string
 	 */
-	public static $newsletter_access_level_meta_name = \Automattic\Jetpack\Extensions\Subscriptions\META_NAME_FOR_POST_LEVEL_ACCESS_SETTINGS;
+	public static $post_access_level_meta_name = \Automattic\Jetpack\Extensions\Subscriptions\META_NAME_FOR_POST_LEVEL_ACCESS_SETTINGS;
```
Comment (romarioraffington) [inline projects/plugins/jetpack/modules/memberships/class-jetpack-memberships.php:43]: Renamed for consistency.

File: projects/plugins/jetpack/modules/memberships/class-jetpack-memberships.php [hunk 5]
```diff
@@ -431,12 +431,26 @@ public static function get_connected_account_id() {
...
 	/**
-	 * Get the newsletter access level
+	 * Get the post access level
 	 *
-	 * @return string|void
+	 * @return string the actual post access level (see projects/plugins/jetpack/extensions/blocks/subscriptions/settings.js for the values).
 	 */
-	public static function get_newsletter_access_level() {
-		return get_post_meta( get_the_ID(), self::$newsletter_access_level_meta_name, true );
+	public static function get_post_access_level() {
```
Comment (romarioraffington) [inline projects/plugins/jetpack/modules/memberships/class-jetpack-memberships.php:439]: Renamed for consistency.

File: projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php [hunk 6]
```diff
@@ -66,7 +66,7 @@ function register_block() {
 	if (
 		/** This filter is documented in class.jetpack-gutenberg.php */
 		! apply_filters( 'jetpack_subscriptions_newsletter_feature_enabled', false )
-		|| ! class_exists( '\Jetpack_Memberships' )
```
Comment (romarioraffington) [inline projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php]: `use Jetpack_Memberships` is in place. No need for `\`.

**Comment by romarioraffington** [comment]
> We should get a WPCOM PR going to address the caching in `earn_get_user_subscriptions_for_site_id` and have some unit tests verify this is working correctly...

Thanks Brent, can you expound on what parts of the caching we need to address here? I see that there is caching in place for `get_user_subscriptions_for_site_id` already 🤔 . Should we create a more robust caching system?

File: projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php (hunk 6 above)
Comment (millerf) [reply projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php]: There is a need for`\`: when used with a string, `class_exists()` needs the full namespace of the class as it checks it in the global context.
There is plenty of literature about that: a quick example [here](https://stackoverflow.com/questions/22407370/how-to-check-if-class-exists-within-a-namespace).
It is not very clear in the official doc, although [this comment](https://www.php.net/manual/en/function.class-exists.php#113581) kind of give it away.

**Comment by n3f** [comment]
> Thanks Brent, can you expound on what parts of the caching we need to address here? I see that there is caching in place for get_user_subscriptions_for_site_id already 🤔 . Should we create a more robust caching system?

See D111411-code

File: projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php (hunk 1 above)
Comment (romarioraffington) [reply projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php]: Fxed in https://github.com/Automattic/jetpack/pull/30868/files

File: projects/plugins/jetpack/extensions/blocks/related-posts/test/fixtures/jetpack__related-posts.json [hunk 7]
```diff
@@ -1,17 +1,17 @@
...
+			"postLayout": "list",
+			"displayDate": false,
+			"displayAuthor": false,
+			"displayThumbnails": true,
+			"displayContext": true,
+			"postsToShow": 4
+		},
+		"innerBlocks": [],
+		"originalContent": ""
+	}
```
Comment (romarioraffington) [inline projects/plugins/jetpack/extensions/blocks/related-posts/test/fixtures/jetpack__related-posts.json:18]: I suspect this was done by the linter

File: projects/plugins/jetpack/extensions/blocks/related-posts/test/fixtures/jetpack__related-posts.parsed.json [hunk 8]
```diff
@@ -1,16 +1,16 @@
...
+			"displayDate": false,
+			"displayAuthor": false,
+			"displayThumbnails": true,
+			"displayContext": true,
+			"postsToShow": 4
+		},
+		"innerBlocks": [],
+		"innerHTML": "",
+		"innerContent": []
+	}
```
Comment (romarioraffington) [inline projects/plugins/jetpack/extensions/blocks/related-posts/test/fixtures/jetpack__related-posts.parsed.json]: I suspect this was done by the linter

**Review by millerf — DISMISSED** [review]

File: projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php (hunk 6 above)
Comment (romarioraffington) [reply projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php]: Re-added in [a147fc0](https://github.com/Automattic/jetpack/pull/30807/commits/a147fc0a3c9f8662a52ad75c866b4271f540b4de)

File: projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php [hunk 9]
```diff
@@ -812,7 +791,7 @@ function get_locked_content_placeholder_text( $newsletter_access_level ) {
 	// Only display this text when Stripe is connected and the post is marked for paid subscribers
 	if (
 		$newsletter_access_level === 'paid_subscribers'
-		&& ! empty( \Jetpack_Memberships::get_connected_account_id() )
+		&& ! empty( Jetpack_Memberships::get_connected_account_id() )
```
Comment (n3f) [inline projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php:808]: Is this change needed?  (I saw @millerf 's comment about the other `\` being needed...)

**Review by n3f — DISMISSED** [review]
WFM -- small question about a backslash -- but it didn't affect my testing...

File: projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php (hunk 9 above)
Comment (romarioraffington) [reply projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php:808]: I could be wrong here but I think his suggestion was related to `class_exists`. For `Jetpack_Memberships` here I added `use Jetpack_Memberships` to the top of this file.

Comment (n3f) [reply projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php:808]: Yup, sounds good.

File: projects/plugins/jetpack/modules/memberships/class-jetpack-memberships.php [hunk 10]
```diff
@@ -432,12 +432,26 @@ public static function get_connected_account_id() {
...
-	 * Get the newsletter access level
+	 * Get the post access level
 	 *
-	 * @return string|void
+	 * @return string the actual post access level (see projects/plugins/jetpack/extensions/blocks/subscriptions/constants.js for the values).
 	 */
-	public static function get_newsletter_access_level() {
-		return get_post_meta( get_the_ID(), self::$newsletter_access_level_meta_name, true );
+	public static function get_post_access_level() {
+		if ( ! is_singular() ) {
```
Comment (romarioraffington) [inline projects/plugins/jetpack/modules/memberships/class-jetpack-memberships.php]: This line is breaking comments gating tests. Checking.

**Comment by millerf** [comment]
> expound

:TIL:

File: projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php [hunk 11]
```diff
@@ -776,8 +768,10 @@ function maybe_close_comments( $default_comments_open, $post_id ) {
 	if ( ! $default_comments_open || ! $post_id ) {
 		return $default_comments_open;
 	}
+
+	$is_post = false;
```
Comment (millerf) [inline projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php]: I am curious about this variable being set here...

I almost thought it was used as a global in `class-jetpack-memberships.php`... It feels a bit misleading.

Also, It looks like `Jetpack_Memberships::user_can_view_post` [does not accept parameters](https://github.com/Automattic/jetpack/pull/30807/files#diff-6c4e5f6a52aca3eb00a4cc6c302da733d70d8d26b61d4ba9e4ec2a8141add6e3R469)... 

Am I missing something here 🤔

File: projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php [hunk 12]
```diff
@@ -478,25 +478,17 @@ function render_block( $attributes ) {
...
  *
- * @return string the actual post access level (see projects/plugins/jetpack/extensions/blocks/subscriptions/settings.js for the values).
+ * @return string the actual post access level (see projects/plugins/jetpack/extensions/blocks/subscriptions/constants.js for the values).
  */
-function get_post_access_level() {
+function get_post_access_level_for_current_post() {
 	if ( ! is_singular() ) {
 		// There is no "actual" current post.
-		return 'everybody';
+		return Token_Subscription_Service::POST_ACCESS_LEVEL_EVERYBODY;
```
Comment (millerf) [inline projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php:488]: Nice!

File: projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php [hunk 13]
```diff
@@ -478,25 +478,17 @@ function render_block( $attributes ) {
 }
 
 /**
- * Generates the source parameter to pass to the iframe
+ *  Get the post access level for the current post. Defaults to 'everybody' if the query is not for a single post
```
Comment (millerf) [inline projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php:481]: Yes, this comment makes more sense. Nice catch!

File: projects/plugins/jetpack/modules/memberships/class-jetpack-memberships.php (hunk 5 above)
Comment (millerf) [reply projects/plugins/jetpack/modules/memberships/class-jetpack-memberships.php:439]: Nice! But also quite hard to read when reviewing (GitHub UI can be tricky)

File: projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php (hunk 6 above)
Comment (millerf) [reply projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php]: Thanks!

**Review by millerf — COMMENTED** [review]
Added a few comments. There are some really nice changes, but I am doubtful about the use of `Jetpack_Memberships::user_can_view_post`...

File: projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php (hunk 11 above)
Comment (romarioraffington) [reply projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php]: Thanks, fixed in [53ed6ba](https://github.com/Automattic/jetpack/pull/30807/commits/53ed6baca5be6d2e6bc078872cbe8b0ba413fe34)

**Review by n3f — APPROVED** [review]
LGTM
